File: rust/src/0003_longest_substring_without_repeating_characters.rs
use std::collections::HashMap;

pub struct Solution;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let (mut longest_length, mut start) = (0, 0);
        let mut map = HashMap::new();

        for (i, val) in s.char_indices() {
            if let Some(&j) = map.get(&val) {
                if j >= start {
                    longest_length = longest_length.max(i - start);
                    start = j + 1;
                }
            }
            map.insert(val, i);
        }

        longest_length.max(s.len() - start) as i32
    }
}

File: 0003_LongestSubstringWithoutRepeatingCharacters.java
import java.util.*;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int start = 0, end = 0;
        int longestLength = 0;
        Set<Character> set = new HashSet<>();

        while (end != s.length()) {
            if (set.contains(s.charAt(end))) {
                set.remove(s.charAt(start));
                start++;
            }
            else {
                set.add(s.charAt(end));
                end++;
                longestLength = Math.max(longestLength, end - start);
            }
        }

        return longestLength;
    }

    public static void main(String[] args) {
        Solution sc = new Solution();
        System.out.println(sc.lengthOfLongestSubstring("abcabcbb"));
    }
}